pkg/pillar/cmd/msrv/pubsub.go

```diff
@@ -457,3 +458,47 @@ func (srv *Msrv) handleAppInstDelete(ctxArg interface{}, key string,
 	srv.unpublishAppInstMetadata(appInstMetadata)
 	srv.Log.Functionf("handleAppInstDelete(%s) done", key)
 }
+
+func (srv *Msrv) lookupAppInstStatusByAppIP(ip net.IP) *types.AppInstanceStatus {
+	sub := srv.subAppInstanceStatus
+	items := sub.GetAll()
+	for _, sc := range items {
+		status := sc.(types.AppInstanceStatus)
+		for _, adapterStatus := range status.AppNetAdapters {
+			if adapterStatus.AllocatedIPv4Addr.Equal(ip) {
+				return &status
+			}
+		}
+	}
+
+	return nil
+}
+
+type AppInstDiscovery struct {
+	Port    uint32 `json:"port"`
+	Address string `json:"network-address"`
+}
+
+func (srv *Msrv) ComposeAppInstancesIPAddresses(UUIDToSkip uuid.UUID) map[string][]AppInstDiscovery {
+	res := make(map[string][]AppInstDiscovery, 0)
+
+	for _, st := range srv.subAppInstanceStatus.GetAll() {
+		status := st.(types.AppInstanceStatus)
+		if status.UUIDandVersion.UUID == UUIDToSkip {
+			continue
+		}
+		adapters := make([]AppInstDiscovery, 0)
+		for _, adapterStatus := range status.AppNetAdapters {
+			if adapterStatus.AllocatedIPv4Addr == nil || adapterStatus.AllocatedIPv4Addr.String() == "" {
+				continue
+			}
+			adapters = append(adapters, AppInstDiscovery{
+				Port:    adapterStatus.IfIdx,
```

Comment:
What does the user do with the interface index?

Reply:
I changed it to Vif. In case where the client application has prior knowledge of the exact physical port on which the server will bind. In this case it will help to find the address client should connect on.